File: go/uapurl/uapurl.go
// Package uapurl extracts the public-facing post/comment URL from a UAP
// metadata blob. Multiple producers (ingest-srv, analysis-srv, knowledge-srv,
// notification-srv) used to maintain their own ordered list of candidate
// keys; centralising the list keeps platform additions in one place.
package uapurl

import "strings"

// CandidateKeys is the ordered set of metadata keys we probe for a URL. The
// first non-empty value wins.
var CandidateKeys = []string{
	"post_url",
	"url",
	"permalink_url",
	"original_url",
	"source_url",
	"web_url",
	"comment_url",
	"share_url",
	"parent_post_url",
}

// FirstFromMap walks CandidateKeys against the given metadata map and returns
// the first non-empty string value, trimmed.
func FirstFromMap(metadata map[string]interface{}) string {
	if len(metadata) == 0 {
		return ""
	}
	for _, key := range CandidateKeys {
		raw, ok := metadata[key]
		if !ok {
			continue
		}
		s, ok := raw.(string)
		if !ok {
			continue
		}
		trimmed := strings.TrimSpace(s)
		if trimmed != "" {
			return trimmed
		}
	}
	return ""
}

File: python/uap_url.py
"""Helpers for extracting a public-facing URL from a UAP metadata dict.

Mirrors shared-libs/go/uapurl so Python services (analysis-srv) and Go
services agree on the candidate key order and the trim semantics.
"""

from __future__ import annotations

from typing import Any, Iterable, Mapping

CANDIDATE_KEYS: tuple[str, ...] = (
    "post_url",
    "url",
    "permalink_url",
    "original_url",
    "source_url",
    "web_url",
    "comment_url",
    "share_url",
    "parent_post_url",
)


def first_from_mapping(metadata: Mapping[str, Any] | None) -> str:
    """Return the first non-empty URL found in *metadata*.

    The check walks CANDIDATE_KEYS in order; any value that is not a string
    or is empty after trim is skipped. Returns ``""`` when nothing matches.
    """

    if not metadata:
        return ""
    for key in CANDIDATE_KEYS:
        value = metadata.get(key)
        if isinstance(value, str):
            trimmed = value.strip()
            if trimmed:
                return trimmed
    return ""


def iter_urls(metadata: Mapping[str, Any] | None) -> Iterable[str]:
    """Yield each non-empty URL in CANDIDATE_KEYS order."""

    if not metadata:
        return
    for key in CANDIDATE_KEYS:
        value = metadata.get(key)
        if isinstance(value, str):
            trimmed = value.strip()
            if trimmed:
                yield trimmed


__all__ = ["CANDIDATE_KEYS", "first_from_mapping", "iter_urls"]
